firstboot: write locale.gen when using --locale if locale-gen is present

DaanDeMeyer opened the pull request:
Probably a bit more controversial than the previous additions but I still think we should do this. Setting the locale on a new system image with --locale isn't exactly useful if the user still has to generate the actual locale himself. 

Questions to answer:
- Do we want to put this behind an option? I'm not entirely sure about other distros, but right now, the locale option isn't exactly all that useful on Arch systems since a user still has to manually run locale-gen to generate the actual locale.
- The current implementation simply uncomments the relevant line in /etc/locale.gen. This assumes that the file has its default format where all supported locales are commented out and the right ones have to be uncommented. Of course, this completely breaks down if /etc/locale.gen doesn't exist or has been modified to remove all the commented lines. However, without the commented lines, I don't really see an easy way of deriving a locale's charset which also has to be specified in /etc/locale.gen. Many but not all locales are suffixed by their charset so using that information wouldn't be fool-proof. On the other hand, systemd-firstboot is meant to be executed on images that haven't been fiddled with so maybe we can make some assumptions on the initial contents of /etc/locale.gen.

Tested with my modified version of mkosi that uses systemd-firstboot and systemd's mkosi Arch configuration.

keszybz commented:
Hmm, is there any reason why one *wouldn't* want to run locale-gen after modifying `/etc/locale.gen`? I'd just make it unconditional, in the sense of running the executable if available.

> This assumes that the file has its default format where all supported locales are commented out and the right ones have to be uncommented. 

That's not nice ;( The general goal is to allow fully empty /etc, so tools should not rely on the presence or contents of any file there when doing "firstboot".

>  I don't really see an easy way of deriving a locale's charset which also has to be specified in /etc/locale.gen

We only need to support UTF-8 encodings as first class citizens. Everything else is obsolete. We can assume that the encoding is UTF-8 when the locale is suffixed with ".UTF-8", so that is easy. And for non-unicode locales we can say "sorry, can't guess the encoding".

keszybz commented on src/firstboot/firstboot.c:
```diff
@@ -1120,6 +1230,12 @@ static int parse_argv(int argc, char *argv[]) {
                                        "--delete-root-password cannot be combined with other root password options");
 
+        if (arg_locale_gen) {
+                r = must_be_root(); /* chroot() */
```
Please don't. In almost all cases must_be_root() is counterproductive. See 569ad251adde02dc0915758fe027e0346e50738a for some discussion.

DaanDeMeyer commented on src/firstboot/firstboot.c:
```diff
@@ -271,12 +273,114 @@ static int prompt_locale(void) {
+static int invoke_locale_gen(void) {
+        int r;
+
+        r = safe_fork("(locale-gen)", FORK_RESET_SIGNALS|FORK_RLIMIT_NOFILE_SAFE|FORK_DEATHSIG|FORK_LOG|FORK_WAIT, NULL);
```
I'm not sure, that would suddenly make systemd-firstboot dependent on a fully booted system, no? I can't immediately think of any major advantages we'd get by using systemd-run here (of course, I might be missing something obvious).

DaanDeMeyer commented:
> Hmm, is there any reason why one wouldn't want to run locale-gen after modifying /etc/locale.gen? I'd just make it unconditional, in the sense of running the executable if available.

After modifying locale.gen, you're basically always going to want to run locale-gen. However, my question was more about whether we want to modify locale.gen and run locale-gen if the --locale option is specified (without adding a --locale-gen option). That would technically be a change in behaviour but on the other hand, configuring a locale in locale.conf that hasn't been generated with locale-gen doesn't seem exactly useful either. I guess there could be a scenario where locales have been pre-generated so you wouldn't have to run locale-gen but even then it doesn't really hurt to run it and I don't think distros currently do this.

The only annoying thing is that with this change, we require root when combining the --root and --locale option since then we need to chroot in order to run locale-gen in the image (although users might be able to get around that by using fakechroot, I haven't tested that).

keszybz commented:
> > Hmm, is there any reason why one wouldn't want to run locale-gen after modifying /etc/locale.gen? I'd just make it unconditional, in the sense of running the executable if available.
> 
> After modifying locale.gen, you're basically always going to want to run locale-gen. However, my question was more about whether we want to modify locale.gen and run locale-gen if the --locale option is specified (without adding a --locale-gen option).

Do you maybe mean the opposite: do we want to run `locale-gen` if `--locale` is **not** specified? It seems obvious to run it if `--locale` is specified and we touch `/etc/locale.gen`.

> That would technically be a change in behaviour but on the other hand, configuring a locale in locale.conf that hasn't been generated with locale-gen doesn't seem exactly useful either. I guess there could be a scenario where locales have been pre-generated so you wouldn't have to run locale-gen but even then it doesn't really hurt to run it and I don't think distros currently do this.

Yeah, I don't think we need to worry too much about this.

> The only annoying thing is that with this change, we require root when combining the --root and --locale option since then we need to chroot in order to run locale-gen in the image (although users might be able to get around that by using fakechroot, I haven't tested that).

Hmm, so yet another idea. What about adding a oneshot service that'd run `locale-gen`, and pulling in that service from/after systemd-firstboot.service?

DaanDeMeyer commented:
> Do you maybe mean the opposite: do we want to run locale-gen if --locale is not specified? It seems obvious to run it if --locale is specified and we touch /etc/locale.gen.

Nah, I meant if --locale is specified, but if it seems obvious to you then I guess its fine to just extend the --locale option to run locale-gen as well. Although there could be some image building scripts using systemd-firstboot that would be running locale-gen twice after we release this (once manually because systemd-firstboot didn't support this yet and once via systemd-firstboot now that it gained support for it). I have a setup script for an Arch install that would be affected by this. Then again, it doesn't hurt to run locale-gen twice and its easy to solve by simply removing the workaround so I'm leaning towards simply extending the --locale option as well.

> Hmm, so yet another idea. What about adding a oneshot service that'd run locale-gen, and pulling in that service from/after systemd-firstboot.service? This service could then be somehow scheduled to run once the image is booted, avoiding the need to run locale-gen from the outside.

@poettering had the same idea in my other PR. However, I'm a little bit worried about tying locale-gen to booting the system. As an example, if we tied generating the locale to booting the system, we'd still have to add a workaround to have a locale generaeted in the mkosi build script image which is run with `systemd-nspawn --as-pid2` so there's no systemd and hence there would be no locale generated (systemd's mkosi build script fails if no UTF-8 locale is available). By generating the locale beforehand, it doesn't matter what a user does with the image, he'll always have a locale available.

DaanDeMeyer commented:
Changed the implementation so it tries to derive the charset from the locale if no commented line can be found in /etc/locale.gen. If we can't derive the locale, we default to UTF-8 as the charset. I also changed stuff to deal with /etc/locale.gen not existing and removed the `must_be_root()` logic.

DaanDeMeyer commented on src/firstboot/firstboot.c:
```diff
@@ -317,6 +318,166 @@ static int process_locale(void) {
+                        if (strncaseeq(locale, l, strlen(locale)))
+                                found = true;
+                        else
+                                l = line;
```
I tried using `extract_first_word` but that broke down with the explanation comments at the start of the file. I'd get en_US.UTF-8 generated twice because it's used as an example in the beginning.

Maybe it's better to leave this logic out and overwrite the entire file instead of trying to work with the original contents? I'll drop the message about being unable to determine the charset to info instead of warning. That should give us a good enough way to determine the charset. Sure we're overwriting the entire file but if people are using the --locale option of systemd-firstboot I don't expect them to be looking at manually modifying the locale again later so maybe we don't gain all that much by trying to work with the original contents of /etc/locale.gen.

DaanDeMeyer replied:
(I dropped the logic for now. Let me know if we should re-add it)

DaanDeMeyer commented on src/firstboot/firstboot.c:
```diff
@@ -317,6 +318,166 @@ static int process_locale(void) {
+                return r;
+
+        r = process_locale_gen();
+        if (r < 0)
```
There's an implicit ordering between process_locale_conf() and process_locale_gen() since process_locale_gen() depends on the prompt_locale() call made by process_locale_conf() so I thought to make that a little bit clearer by putting them in a helper function.

I'm not particularly convinced by that approach though so I can remove it if you think it doesn't add any value. A comment might provide the same utility.

DaanDeMeyer replied:
I removed the helper and added a comment instead.

DaanDeMeyer commented on src/firstboot/firstboot.c:
```diff
@@ -317,6 +318,166 @@ static int process_locale(void) {
+        const char *etc_localegen;
+        int r;
+
+        r = find_binary("locale-gen", NULL);
```
I removed the `find_binary` check since I don't think we can call it after a fork since it allocates memory. I guess we can just fail if we don't find it? Unless we want to support doing the locale stuff on non-glibc systems. I'm not sure if that's supported.

DaanDeMeyer commented on src/firstboot/firstboot.c:
```diff
@@ -317,6 +318,96 @@ static int process_locale(void) {
+                                log_error_errno(errno, "Failed to chroot to %s: %m", arg_root);
+                                _exit(EXIT_FAILURE);
+                        }
+                }
```
Actually, even if we did that, I have no idea if the locale-archive generated by the host locale-gen (actually localedef underneath) would be compatible with whatever glibc version that's running in the image so adding the --root option might not even work at all.

Honestly, the whole concept of generating locales like this seems outdated. Why not simply allow installing locales like any other package? That's not systemd's problem to solve though.

I'll leave the service for another PR. I'm still not entirely convinced its the way to go.

DaanDeMeyer commented on src/firstboot/firstboot.c:
```diff
@@ -317,6 +318,96 @@ static int process_locale(void) {
+        if (r < 0)
+                return -ENOMEM;
+
+        r = write_string_file(etc_localegen, line,
```
I added the auto-detect but I'm not sure if we should make this optional. I feel its generally gonna make sense to do this if locale-gen is available. Maybe wait until someone complains before adding an option?

poettering commented on src/basic/path-util.c:
```diff
@@ -633,7 +646,9 @@ int find_binary(const char *name, char **ret) {
                         continue;
 
-                j = path_join(element, name);
+                path = prefix_roota(root, element);
```
this allocates on the stack, and that's not OK in a loop of any kind

poettering commented on src/basic/path-util.c:
```diff
@@ -633,7 +646,9 @@ int find_binary(const char *name, char **ret) {
-                j = path_join(element, name);
+                path = prefix_roota(root, element);
+
+                j = path_join(path, name);
```
so, the access() further down will resolve symlinks relative to the current mount namespace. Which is wrong. You need to use chase_symlinks() i guess, to find the binary, so that symlinks are resolved relative to the specified root.

DaanDeMeyer commented:
Seems github somehow gave you an outdated version of the PR to review :(. 

I'm not entirely sure if this is the way to go anymore though. Especially since we might be getting upstream C.UTF-8 support soon (https://sourceware.org/pipermail/libc-alpha/2020-June/115290.html) which would remove my primary motivation for adding this in the first place (getting a UTF-8 locale in container images). I'll close this for now until we see where that goes.
